text/0006-BundleContext-Install-API.md

```diff
@@ -0,0 +1,136 @@
+*       by this point, the headers are correct.
+*
+* @param manifestHeaders a map of the headers for a bundle manifest.
+*/
+explicit BundleManifest::BundleManifest(const AnyMap& manifestHeaders);
+
+```
+
+### BundlePrivate
+
+* Additional constructor that takes bundle manifest to be injected
+* Modified to fetch bundle manifest from BundleArchive instead of parsing the resource
+* Factor out manifest validation from constructor to allow for use from new constructor which takes a manifest to be injected as an argument
+
+### BundleStorage
+
+The **BundleStorage** class hierarchy is used for implementing different strategies for tracking bundle **BundleArchives**. There are two subclasses, but only one of them provides an actual implementation. They are **BundleStorageMemory**, and **BundleStorageFile**. It also tracks which bundles are to be auto started.
+
+```c++
+  /**
+   * Insert bundles from a container into persistent storagedata.
+   *
+   * This is a simpler version of the InstallBundles api that only deals with one bundle at a time.
+   * 
+   * @param resCont The container for the bundle data and resources.
+   * @param topLevelEntries The top level entries in the container to be inserted as bundle archives.
+   * @return A list of BundleArchive instances representing the installed bundles.
+   */
+  using ManifestT = cppmicroservices::AnyMap;
+  virtual std::shared_ptr<BundleArchive> InsertArchive(const  std::shared_ptr<BundleResourceContainer>& resCont
+                                                       , const std::string& topLevelEntry
+                                                       , const ManifestT&) = 0;
+
+```
+
+* Combined BundleStorage::InsertBundleLib and BundleStorage::InsertArchives
+  * InsertBundleLib was a wrapper around InsertArchives
+    * It created a BundleResourceContainer for the location and called InsertArchives
+    * Now the BundleResourceContainer is created if needed in BundleRegistry::GetAlreadyInstalldBundlesAtLocation and then passed in to BundleStorage::InsertArchive.
+
+### BundleArchive
+
+* Refactor and remove Data class simplifying design.
+* Add BundleManifest and constructor argument
+  * Call BundleManifest::Parse from BundleArchive in the event that a manifest is not passed in at construction
+* Add new **GetManifest()** method which does a lazy loadof the manifest from the archive if and only if we don't already have the manifest.
+
+### BundleResourceContainer
+
+* Add **manifest** argument to constructor
+  * If the **manifest** is not empty, use it to initialize the **m_SortedToplevelDirs**, and leave **m_IsContainerOpen** set to false
+    * if **manifest** is empty, no manifest injection is happening, so open up the archive and read the manifest from there, set set **m_IsContainerOpen** to true.
+
+### BundleStorage(Memory)
+
+* Refactor
+  * Remove **InsertArchives()** method and move loop to caller in **BundleRegistry::Install0**
+
```

Comment:
a sequence diagram showing how the abovementione classes now interact when `InstallBundles` is called would be helpful to illustrate the proposed changes.

Reply:
Done.